stone/frontend/ir_generator.py

```diff
@@ -1248,3 +1259,235 @@ def mk_route_schema():
+                data_types, routes_by_ns = self._get_data_types_and_routes_from_doc_ref(
+                    namespace.doc, namespace_name)
+                for d in data_types:
+                    route_data_types.append(d)
+                for ns_name, routes in routes_by_ns.items():
+                    ns = self.api.namespaces[ns_name]
+                    for r in routes:
+                        for d in ns.get_route_io_data_types_for_route(r):
+                            route_data_types.append(d)
+
+            # Parse user-specified routes and add them to the starting data types
+            # Note that this may add duplicates, but that's okay, as the recursion
+            # keeps track of visited data types.
+            if routes_names == ['*']:
+                routes_names = namespace.route_by_name.keys()
+            #     route_data_types.extend(namespace.get_route_io_data_types())
+            # else:
+            assert '*' not in routes_names
+            for route_name in routes_names:
+                if route_name not in namespace.route_by_name:
+                    print('Route %s is not defined!' % route_name, file=sys.stderr)
+                    sys.exit(1)
+                route = namespace.route_by_name[route_name]
+                route_data_types.extend(namespace.get_route_io_data_types_for_route(route))
+                if route.doc is not None:
+                    data_types, routes_by_ns = self._get_data_types_and_routes_from_doc_ref(
+                        route.doc, namespace_name)
+                    for d in data_types:
+                        route_data_types.append(d)
+                    for ns_name, routes in routes_by_ns.items():
+                        ns = self.api.namespaces[ns_name]
+                        for r in routes:
+                            for d in ns.get_route_io_data_types_for_route(r):
+                                route_data_types.append(d)
+
+        # Parse the datatype whitelist and populate any starting data types
+        for namespace_name, datatype_names in self._routes['datatype_whitelist'].items():
+            if namespace_name not in self.api.namespaces:
+                print('Namespace %s is not defined!' % namespace_name, file=sys.stderr)
+                sys.exit(1)
+
+            # Parse namespace doc refs and add them to the starting data types
+            namespace = self.api.namespaces[namespace_name]
+            if namespace.doc is not None:
+                data_types, routes_by_ns = self._get_data_types_and_routes_from_doc_ref(
+                    namespace.doc, namespace.name)
+                for d in data_types:
+                    route_data_types.append(d)
+                for ns_name, routes in routes_by_ns.items():
+                    ns = self.api.namespaces[ns_name]
+                    for r in routes:
+                        for d in ns.get_route_io_data_types_for_route(r):
+                            route_data_types.append(d)
```

Comment:
This pattern in lines 1327 - 1335 shows up three different times in this method, is there a way to DRY this logic so you don't have the same code copied three times?